### PR TITLE
fix consensus

### PR DESCRIPTION
**Pull request by erikzhang**

### CODE REVIEW

**Comment by vncoelho** [comment]
@erikzhang, coincidentally, we have been also working on this today at #426.
I think that we did a nice design that should be considered.

**Comment by vncoelho** [comment]
This fix is more related to avoiding communication after block is sent, not so much related to the commit phase we have been working, right? :dagger:

**Comment by shargon** [comment]
@erikzhang could you explain what happend here? why the necesity of review the signatures?

https://github.com/neo-project/neo/blob/92e55f13fd120c5cdd15ff46dcccd0a205b2ce33/neo/Consensus/ConsensusService.cs#L252-L257

**Comment by vncoelho** [comment]
What if a normal RPC inject this, Shargon?

**Comment by igormcoelho** [comment]
I think  its not reviewing signatures shargon, its allowing to.keep sigs from PrepReponse if PrepRequest arrives later.. the previous code was dropping useful signatures I guess.

**Comment by erikzhang** [comment]
> I think its not reviewing signatures shargon, its allowing to.keep sigs from PrepReponse if PrepRequest arrives later.. the previous code was dropping useful signatures I guess.

Yes

**Comment by shargon** [comment]
It seems good for me, we will test it! :)

**Comment by erikzhang** [comment]
I updated it again and solved the consensus problem that is now facing. Please review it again. @shargon @vncoelho @igormcoelho

**Comment by vncoelho** [comment]
give me 2 hours, Erik, I was drafting something about this during my bus trip...ahauaha

**Comment by vncoelho** [comment]
Double checked, @erikzhang. Great job and precise.

---------------------------------------------------------------------------------------------------
Today I have been also spent the day reflecting about this verify:
https://github.com/neo-project/neo/blob/c2d7bd44142e3740b878c8feb86ac05d930cf5c5/neo/Network/P2P/Payloads/Transaction.cs#L253
This aforementioned line is called when the Backup does not have a given hash and asks for it (`OnTransaction`). In this case, the transaction is also verified.
https://github.com/neo-project/neo/blob/c2d7bd44142e3740b878c8feb86ac05d930cf5c5/neo/Consensus/ConsensusService.cs#L346

My doubt was related to the order that it is verified, because it is verified against `Snapshot snapshot, IEnumerable<Transaction> mempool`, which is `context.Transactions.Values`.
https://github.com/neo-project/neo/blob/c2d7bd44142e3740b878c8feb86ac05d930cf5c5/neo/Consensus/ConsensusService.cs#L38
Thus, I was thinking if the Verify should consider the order that this transaction is included or it does not matter? Because some time ago we were discussing about how Storage states changes according to the order that Transactions are processed sequentially.
----------------------------------------------------------------------------------------------------------

**Comment by vncoelho** [comment]
Erik, what if you verify the transactions `OnPersist` and, then, delete/remove the one that fails the test along with these ones here https://github.com/neo-project/neo/blob/c2d7bd44142e3740b878c8feb86ac05d930cf5c5/neo/Ledger/Blockchain.cs#L382?

**Comment by igormcoelho** [comment]
@erikzhang This change just feels redundant to me, because these tx are supposed to exist on mempool already... so perhaps it's something I didn't get. The most important is to get Neo safer and safer now, even if it costs performance temporarialy.

**Comment by shargon** [comment]
@erikzhang if a CN doesn't have a TX, where he asking for it?

**Comment by erikzhang** [comment]
> Erik, what if you verify the transactions OnPersist and, then, delete/remove the ones that fails the test along with these ones here

If there are a large number of transactions in the memory pool, it takes a lot of time to re-verify them all.

> This change just feels redundant to me, because these tx are supposed to exist on mempool already...

After the block is persisted, all transactions in the memory pool must be re-verified.

> if a CN doesn't have a TX, where he asking for it?

https://github.com/neo-project/neo/blob/c2d7bd44142e3740b878c8feb86ac05d930cf5c5/neo/Consensus/ConsensusService.cs#L325-L330

The `primary` will re-relay the transactions after sending the `PrepareRequest`.

**Comment by vncoelho** [comment]
Thanks, Erik.

>If there are a large number of transactions in the memory pool, it takes a lot of time to re-verify them all.

Does this re-verifiying occurs naturally at the `OnNewTransaction` events (however, depends on P2P communication)?

**One last question and summary, Erik. :dagger:**
1. `OnPersistCompleted`  `Self.Tell(tx, ActorRefs.NoSender)` send a message to All Connections with all current reimaining TXs in the mempool, right?
1. mem_pool is clear;
1. Block is distributed to everyone again;
1. Speaker FillContext only using memory pool `.GetMemoryPool()`. Since mem_pool was clear, Speaker memory pool just contains  transactions that arrived though `OnNewTransaction` calls?
1. Backups checks the `Primary Context` and get missing transactions from `mem_pool_unverified` for speed up instead of waiting for a transaction to arrive naturally as `OnNewTransaction`.

Thanks for the patient, Erik.

Following my reasoning I would see this `mem_pool_unverified` as a speed up instead of a fix.

**Comment by belane** [comment]
Tested on private net and working fine.
```
==> /opt/node1/neo-cli/Logs/2018-10-26.log <==
[16:06:12.316] initialize: height=90 view=0 index=2 role=Primary
[16:06:27.323] timeout: height=90 view=0 state=Primary
[16:06:27.324] send prepare request: height=90 view=0
[16:06:27.345] OnPrepareResponseReceived: height=90 view=0 index=3
[16:06:27.380] OnPrepareResponseReceived: height=90 view=0 index=3
[16:06:27.385] OnPrepareResponseReceived: height=90 view=0 index=3
[16:06:27.390] OnPrepareResponseReceived: height=90 view=0 index=0
[16:06:27.394] relay block: 0x5440aad0193a621d92555f35b914b5e1f71743d1c94b8b94df64b69161308d4a
[16:06:27.397] OnPrepareResponseReceived: height=90 view=0 index=0
[16:06:27.400] OnPrepareResponseReceived: height=90 view=0 index=0
[16:06:27.401] OnPrepareResponseReceived: height=90 view=0 index=1
[16:06:27.405] OnPrepareResponseReceived: height=90 view=0 index=1
[16:06:27.410] OnPrepareResponseReceived: height=90 view=0 index=1
[16:06:27.413] persist block: 0x5440aad0193a621d92555f35b914b5e1f71743d1c94b8b94df64b69161308d4a

==> /opt/node2/neo-cli/Logs/2018-10-26.log <==
[16:06:12.322] initialize: height=90 view=0 index=0 role=Backup
[16:06:27.328] OnPrepareRequestReceived: height=90 view=0 index=2 tx=1
[16:06:27.340] send prepare response
[16:06:27.341] OnPrepareRequestReceived: height=90 view=0 index=2 tx=1
[16:06:27.345] OnPrepareRequestReceived: height=90 view=0 index=2 tx=1
[16:06:27.346] OnPrepareResponseReceived: height=90 view=0 index=3
[16:06:27.352] relay block: 0x5440aad0193a621d92555f35b914b5e1f71743d1c94b8b94df64b69161308d4a
[16:06:27.358] OnPrepareResponseReceived: height=90 view=0 index=3
[16:06:27.360] OnPrepareResponseReceived: height=90 view=0 index=3
[16:06:27.361] OnPrepareResponseReceived: height=90 view=0 index=1
[16:06:27.365] OnPrepareResponseReceived: height=90 view=0 index=1
[16:06:27.368] OnPrepareResponseReceived: height=90 view=0 index=1
[16:06:27.376] persist block: 0x5440aad0193a621d92555f35b914b5e1f71743d1c94b8b94df64b69161308d4a

==> /opt/node3/neo-cli/Logs/2018-10-26.log <==
[16:06:12.337] initialize: height=90 view=0 index=3 role=Backup
[16:06:27.328] OnPrepareRequestReceived: height=90 view=0 index=2 tx=1
[16:06:27.334] send prepare response
[16:06:27.335] OnPrepareRequestReceived: height=90 view=0 index=2 tx=1
[16:06:27.335] OnPrepareRequestReceived: height=90 view=0 index=2 tx=1
[16:06:27.342] OnPrepareResponseReceived: height=90 view=0 index=0
[16:06:27.364] relay block: 0x5440aad0193a621d92555f35b914b5e1f71743d1c94b8b94df64b69161308d4a
[16:06:27.373] OnPrepareResponseReceived: height=90 view=0 index=0
[16:06:27.374] OnPrepareResponseReceived: height=90 view=0 index=0
[16:06:27.374] OnPrepareResponseReceived: height=90 view=0 index=1
[16:06:27.375] OnPrepareResponseReceived: height=90 view=0 index=1
[16:06:27.386] OnPrepareResponseReceived: height=90 view=0 index=1
[16:06:27.398] persist block: 0x5440aad0193a621d92555f35b914b5e1f71743d1c94b8b94df64b69161308d4a

==> /opt/node4/neo-cli/Logs/2018-10-26.log <==
[16:06:12.325] initialize: height=90 view=0 index=1 role=Backup
[16:06:27.335] OnPrepareRequestReceived: height=90 view=0 index=2 tx=1
[16:06:27.343] send prepare response
[16:06:27.347] OnPrepareRequestReceived: height=90 view=0 index=2 tx=1
[16:06:27.347] OnPrepareRequestReceived: height=90 view=0 index=2 tx=1
[16:06:27.348] OnPrepareResponseReceived: height=90 view=0 index=3
[16:06:27.355] relay block: 0x5440aad0193a621d92555f35b914b5e1f71743d1c94b8b94df64b69161308d4a
[16:06:27.369] OnPrepareResponseReceived: height=90 view=0 index=3
[16:06:27.371] OnPrepareResponseReceived: height=90 view=0 index=3
[16:06:27.372] OnPrepareResponseReceived: height=90 view=0 index=0
[16:06:27.380] OnPrepareResponseReceived: height=90 view=0 index=0
[16:06:27.383] OnPrepareResponseReceived: height=90 view=0 index=0
[16:06:27.388] persist block: 0x5440aad0193a621d92555f35b914b5e1f71743d1c94b8b94df64b69161308d4a
```

**Comment by belane** [comment]
I would prefer to re-verify the pending TXs in `Blockchain.OnPersistCompleted` insted to move them to `mem_pool_unverified` because they are now available only for backup nodes if they receive a `PrepareRequest` with one of these TX but are not available for the Primary in `FillContext`.

**Comment by shargon** [comment]
This lock the state when a Block was sent, right? do you think that is needed the commit phase after this PR?

**Comment by vncoelho** [comment]
@erikzhang, I agree with you that we should not re-verify at OnPersist.

I am trying to understand how the mempool is filled again. All transaction that were cleared arrive now with `Self.Tell(tx, ActorRefs.NoSender)` as a `OnNewTransaction` event?

**Comment by erikzhang** [comment]
> I think that this mem_pool_unverified is a speed up and not a fix, right?

Right.

**Comment by vncoelho** [comment]
Lets go for it and then we can merge with the commit phase for finally tackling that.

File: neo/Consensus/ConsensusService.cs [hunk 1]
```diff
@@ -272,12 +289,17 @@ private void OnPrepareRequestReceived(ConsensusPayload payload, PrepareRequest m
...
-            if (context.State.HasFlag(ConsensusState.BlockSent)) return;
             if (context.Signatures[payload.ValidatorIndex] != null) return;
-            Block header = context.MakeHeader();
-            if (header == null || !Crypto.Default.VerifySignature(header.GetHashData(), message.Signature, context.Validators[payload.ValidatorIndex].EncodePoint(false))) return;
-            context.Signatures[payload.ValidatorIndex] = message.Signature;
-            CheckSignatures();
+            byte[] hashData = context.MakeHeader()?.GetHashData();
+            if (hashData == null)
+            {
+                context.Signatures[payload.ValidatorIndex] = message.Signature;
```
Comment (vncoelho) [inline neo/Consensus/ConsensusService.cs]: @erikzhang, I am double checking here and just to be sure I am thinking why `message.Signature` is assigned here to be teh Signatures array.

Should not it be `context.Signatures[payload.ValidatorIndex] = null`?
If the hashData is null it means that the payload is kind of empty, right?

**Comment by erikzhang** [comment]
I think @igormcoelho has answered your question:

> I think its not reviewing signatures shargon, its allowing to.keep sigs from PrepReponse if PrepRequest arrives later.. the previous code was dropping useful signatures I guess.

**Pull request merged**